retrieval/target_provider.go

```diff
@@ -117,19 +161,77 @@ func (p *sdTargetProvider) Targets() ([]Target, error) {
 		if addr.Target[len(addr.Target)-1] == '.' {
 			addr.Target = addr.Target[:len(addr.Target)-1]
 		}
-		endpoint.Host = fmt.Sprintf("%s:%d", addr.Target, addr.Port)
-		t := NewTarget(endpoint.String(), p.job.ScrapeTimeout(), baseLabels)
+		tp.endpoint.Host = fmt.Sprintf("%s:%d", addr.Target, addr.Port)
+		t := NewTarget(tp.endpoint.String(), tp.deadline, tp.baseLabels)
 		targets = append(targets, t)
 	}
+	return targets, err
+}
 
-	p.targets = targets
+// etcdTargetProvider provides target information from the etcd key-value store
+// based on a provided config file.
+type etcdTargetProvider struct {
+	srv        config.ServiceDiscoveryConfig
+	endpoint   *url.URL
+	baseLabels clientmodel.LabelSet
+	deadline   time.Duration
+}
+
+// Targets implements the TargetProvider interface.
+func (tp *etcdTargetProvider) Targets() ([]Target, error) {
+	cfgFile := tp.srv.GetConfigFile()
+	if cfgFile == "" {
+		return nil, fmt.Errorf("missing config file for etcd service discovery")
+	}
+	c, err := etcd.NewClientFromFile(cfgFile)
+	if err != nil {
+		return nil, fmt.Errorf("etcd discovery: error creating client: %s", err)
+	}
+	resp, err := c.Get(tp.srv.GetName(), true, true)
+	if err != nil {
+		return nil, fmt.Errorf("etcd discovery: error on get request: %s", err)
+	}
+	// we expect the service information to be located in the node itself
```

Comment:
No, certainly not (even though that's probably how most people would do it.)
Being able to provide a list instead of a single key would be a first step to be more flexible. Also visiting all nodes below the given key.

People could store a comma separated list in a single field, too - but we probably should stop somewhere.